File: examples/C/read_and_write.c
#include <stdlib.h>
#include <stdio.h>
#include "jhpcndf.h"

#define NUM_DATA 10000

#ifdef _REAL_IS_DOUBLE_
#define REAL_TYPE double
#else
#define REAL_TYPE float
#endif


int main(int argc, char *argv[])
{

    //
    // JHPCN-DF Write example
    //

    // SetUp Data
    const float tolerance=0.01;
    const int num_data=NUM_DATA;
    REAL_TYPE* random_data= malloc(num_data*sizeof(REAL_TYPE));
    for(int i=0; i<num_data; i++)
    {
        random_data[i]=rand()/((REAL_TYPE)(RAND_MAX)+1);
    }

    // Open file
    int key=JHPCNDF_fopen("upper_bits", "lower_bits", "w+b", "gzip", 32768);

    // Encode and write
#ifdef _REAL_IS_DOUBLE_
    JHPCNDF_fwrite_double(random_data, sizeof(REAL_TYPE), num_data, key, tolerance, 1, "binary_search");
#else
    JHPCNDF_fwrite_float(random_data, sizeof(REAL_TYPE), num_data, key, tolerance, 1, "binary_search");
#endif

    // Close file
    JHPCNDF_fclose(key);


    //
    // JHPCN-DF Read example
    //

    REAL_TYPE* work = malloc(num_data*sizeof(REAL_TYPE));

    // OpenFile
    int key2=JHPCNDF_fopen("upper_bits", "lower_bits", "rb", "gzip", 32768);

    // Read and decode
#ifdef _REAL_IS_DOUBLE_
    JHPCNDF_fread_double(work, sizeof(REAL_TYPE), num_data, key2);
#else
    JHPCNDF_fread_float(work, sizeof(REAL_TYPE), num_data, key2);
#endif

    // Close file
    JHPCNDF_fclose(key2);


    // compare files
    int error_count=0;
    for(int i=0; i<num_data; i++)
    {
        if (random_data[i] != work[i])
        {
            if(error_count <10)
            {
                fprintf(stderr, "%d th data is not same.\n",i);
            }
            error_count++;
        }
    }
    if (error_count >0)
    {
        fprintf(stderr, "%d entry is not correctly restored.\n", error_count);
    }else{
        fprintf(stderr, "%d data is correctly write and read.\n", num_data);
    }

    return 0;
}

File: examples/Fortran/read_and_write.f90
program f_sample
use jhpcndf
implicit none
integer(8),parameter :: num_data=10000
integer(4)           :: i, error_count, key, key2
real(4)              :: tolerance
integer(8),parameter :: buff=32768

! FREALOPTによりr4/r8を切り替え
real, allocatable  :: random_data(:)
real, allocatable  :: work(:)

! setup
tolerance=0.01
allocate(random_data(num_data))
call random_number(random_data)



!
! JHPCN-DF Write example
!

! open file
call JHPCNDF_OPEN(UNIT=KEY, UFILE='upper_bits', LFILE='lower_bits', ACTION='WRITE', comp="gzip", buff_size=buff)


! encode and write
call JHPCNDF_WRITE(UNIT=KEY, RECL=num_data, data=random_data, TOL=tolerance, IS_REL=.true., ENC="binary_search")

! close file
call JHPCNDF_CLOSE(KEY)
 


!
! JHPCN-DF Read example
!
allocate(work(num_data))

! open file
call JHPCNDF_OPEN(UNIT=KEY2, UFILE='upper_bits', LFILE='lower_bits', ACTION='READ', comp="gzip", buff_size=buff)

! read file and decode
call JHPCNDF_READ(UNIT=KEY2, RECL=num_data, data=work)

! close file
call JHPCNDF_CLOSE(KEY2)

! compare original and read data
error_count=0
do i=1, num_data
if (random_data(i) .ne. work(i)) then 
    if (error_count .lt. 10) then
        write(0,*) i,' th entry is not same:', random_data(i), work(i)
    endif
    error_count=error_count+1
endif
enddo


if (error_count .gt. 0) then
    write(0,*) error_count,' entry is not correctly restored.' 
else
    write(0,*) num_data,' data is correctly write and read.'
endif


end program
